test/parallel/test-crypto-key-objects-messageport.js

```diff
@@ -0,0 +1,55 @@
+'use strict';
+const common = require('../common');
+if (!common.hasCrypto)
+  common.skip('missing crypto');
+
+const assert = require('assert');
+const { createSecretKey, generateKeyPairSync } = require('crypto');
+const { createContext } = require('vm');
+const { MessageChannel, moveMessagePortToContext } = require('worker_threads');
+
+const secretKey = createSecretKey(Buffer.alloc(32));
+const { publicKey, privateKey } = generateKeyPairSync('rsa', {
+  modulusLength: 1024
+});
+
+function keyToString(key) {
+  if (key.type === 'secret') {
+    return key.export().toString('hex');
+  } else {
+    return key.export({ type: 'pkcs1', format: 'pem' });
+  }
+}
+
+// Get immutable representations of all keys.
+const keys = [secretKey, publicKey, privateKey]
+             .map((key) => [key, keyToString(key)]);
+
+for (const [key, repr] of keys) {
+  {
+    const { port1, port2 } = new MessageChannel();
+
+    port1.postMessage({ key });
+    assert.strictEqual(keyToString(key), repr);
+
+    port2.once('message', common.mustCall(({ key }) => {
+      assert.strictEqual(keyToString(key), repr);
+    }));
+  }
+
+  {
+    // Test 3: Clone a dummy object to a message port in another context.
+    const { port1, port2 } = new MessageChannel();
+    const context = createContext();
+    const port2Moved = moveMessagePortToContext(port2, context);
+    assert(!(port2Moved instanceof Object));
+
+    port2Moved.onmessage = common.mustCall(({ data: { key } }) => {
+      assert.strictEqual(keyToString(key), repr);
```

Comment:
I assume that `key instanceof Object` is **true** here, but `port2Moved` should only create objects in `context` and not in the outer/main context (see also the comment for `TransferData::Deserialize()`).

`NativeKeyObject::KeyObjectTransferData::Deserialize()` should either fail if `context != env->context()`, or `env->crypto_key_object_secret_constructor()` etc. need to be stored as templates instead.

Reply:
Can you give me a hint how to store the constructors as templates? I didn't think it was possible with classes defined in JavaScript :)

Reply:
Ah right, it’s not … I guess that means that you’ll have to throw an exception here if the contexts mismatch

(The alternative would be giving node_crypto multi-context support, but we’re not really quite there yet I’d say … and in any case, that would be a lot more work than this PR)

Reply:
I can't quite figure out where to throw. Throwing in `Deserialize` compiles, but then how would the user catch it? It doesn't seem to cause an `error` event. Throwing in `CloneForMessaging` seems like the better choice to me, but that seems to be even more difficult (since it only happens in `SerializerDelegate::Finish` without access to the environment/context etc.). Do you have another hint for me? :)

Reply:
Hm – `Deserialize()` would be the right place, yes. Maybe adding

```c++
  TryCatchScope try_catch(env());
  try_catch.SetVerbose(true);
```

after the `Context::Scope context_scope(context);` line in `MessagePort::OnMessage()` helps? It would still result in an uncaught exception, I think, but I’m not really sure what the alternative would be… there *is* `onmessageerror` on the Web `MessagePort`, but we’re not currently implementing that.

(I think even just discarding the message would also be fine as a step for now – really anything that does not result in an object being created in the wrong `Context`.)